numcodecs/shuffle.py

```diff
@@ -0,0 +1,109 @@
+
+# The _doShuffle and _doUnshuffle functions have been lifted from HSDS which
+# includes the following notice:
+##############################################################################
+# Copyright by The HDF Group.                                                #
+# All rights reserved.                                                       #
+#                                                                            #
+# This file is part of HSDS (HDF5 Scalable Data Service), Libraries and      #
+# Utilities.  The full HSDS copyright notice, including                      #
+# terms governing use, modification, and redistribution, is contained in     #
+# the file COPYING, which can be found at the root of the source code        #
+# distribution tree.  If you do not have access to this file, you may        #
+# request a copy from help@hdfgroup.org.                                     #
+##############################################################################
+
+from numba import jit
+import numpy as np
+from .compat import ensure_bytes, ensure_contiguous_ndarray, ndarray_copy, ensure_ndarray
+from .abc import Codec
+
+
+@jit(nopython=True)
+def _doShuffle(src, des, element_size):
+    count = len(src) // element_size
+    for i in range(count):
+        offset = i*element_size
+        e = src[offset:(offset+element_size)]
+        for byte_index in range(element_size):
+            j = byte_index*count + i
+            des[j] = e[byte_index]
+    return des
+
+
+@jit(nopython=True)
+def _doUnshuffle(src, des, element_size):
+    count = len(src) // element_size
+    for i in range(element_size):
+        offset = i*count
+        e = src[offset:(offset+count)]
+        for byte_index in range(count):
+            j = byte_index*element_size + i
+            des[j] = e[byte_index]
+    return des
```

Comment:
Does NumPy have some mechanism for doing this directly?

Reply:
Probably using some indexing operations, however I expect it would be much slower, but haven't benchmarked